hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/utils/TransactionUtils.java

```diff
@@ -71,12 +71,12 @@ public static Option<HoodieCommitMetadata> resolveWriteConflictIfAny(
     WriteOperationType operationType = thisCommitMetadata.map(HoodieCommitMetadata::getOperationType).orElse(null);
     if (config.needResolveWriteConflict(operationType)) {
       // deal with pendingInstants
-      Stream<HoodieInstant> completedInstantsDuringCurrentWriteOperation = getCompletedInstantsDuringCurrentWriteOperation(table.getMetaClient(), pendingInstants);
-
-      ConflictResolutionStrategy resolutionStrategy = config.getWriteConflictResolutionStrategy();
       if (reloadActiveTimeline) {
         table.getMetaClient().reloadActiveTimeline();
       }
+      Stream<HoodieInstant> completedInstantsDuringCurrentWriteOperation = getCompletedInstantsDuringCurrentWriteOperation(table.getMetaClient(), pendingInstants);
```

Comment:
It looks like `getCompletedInstantsDuringCurrentWriteOperation` always need a timeline reload, maybe we should just remove line 74 ~ line 76

Reply:
why does it always require a reload?

Reply:
Because a concurrent writer may already complete some pending instants(from current writer's insight).

Reply:
Yes, as Vinoth has pointed out this is not required because it is reloaded within the lock already. That is why there is a parameter on the method.

Reply:
Can you point out the line that "reload the timeline within the lock", if it is already realoded, what is the purpose for another reload?

Reply:
There is an argument to this method called `reloadActiveTimeline` and it is true [here](https://github.com/apache/hudi/blob/master/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/commit/BaseCommitActionExecutor.java#L202-L203) since the table is created outside of the lock.

It is false [here](https://github.com/apache/hudi/blob/master/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieClient.java#L239-L240). This method is called from the preCommit methods in [BaseHoodieWriteClient](https://github.com/apache/hudi/blob/master/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieWriteClient.java#L378) and [BaseHoodieTableServiceClient](https://github.com/apache/hudi/blob/master/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieTableServiceClient.java#L179). Both of these are creating the table instance which will create a new meta client right before calling the method linked in `BaseHoodieWriteClient`

Reply:
Thanks for the elaboration, after this change, we need to be cautious for the parameter `reloadActiveTimeline` so that the call `getCompletedInstantsDuringCurrentWriteOperation` can always check based on the latest state .

Reply:
We could also rename this variable so it's clear it is related to whether the timeline was refreshed within the lock

Reply:
yeah, let's give it a better name.

Reply:
How about `timelineInitializedOutsideOfTransaction` or something along those lines?

Reply:
maybe `timelineRefreshed` to indicate whether the timeline is already the latest while invoking this method.

Reply:
I was hoping to have some context about the transaction in the name so it is clear to future readers. I will update the javadocs as well.

Reply:
@danny0405 how about `timelineRefreshedWithinTransaction`?

Reply:
I'm okay with it.

Reply:
Updated the PR